introduction_and_learning/Histogram/histograms.py

```diff
@@ -0,0 +1,291 @@
+__all__ = [
+    "HistogramTxt",
+    "HistogramJSON",
+    "HistogramXml"
+]
+
+from histogram_api import HistogramInterface
+import json
+import xml.etree.ElementTree as ET
+import csv
+
+
+class HistogramTxt(HistogramInterface):
+    def read_data(self, file_path: str) -> str:
+        extension = file_path[file_path.rfind("."):]
+
+        if extension == ".txt":
+            with open(file_path, "tr", encoding="utf8") as f:
+                return f.read()
+        else:
+            return "File does not have a .txt extension."
+
+    def create(self, data_input: str) -> dict:
+        # Note: Delete from string characters "." and "," then transform in to
+        # ordered list.
+        words = data_input.casefold().replace(".", "").replace(",", "").split(" ")
+        words.sort()
+        # --- end note
+        hist = {}
+
+        for word in words:
+            if word in hist:
+                hist[word] += 1
+            else:
+                hist[word] = 1
+
+        return hist
+
+    def generate_report(self, histogram: dict, output_type: str = "html", path: str = "report") -> bool:
+        output_types = ["html", "csv", "txt"]
+
+        if output_type not in output_types:
+            return False
+        else:
+            if output_type == "html":
+                output = ["<html>\n<body>\n\t<table>\n\t<tr><th>Name</th>"
+                          "<th>Amount</th></tr>"]
+
+                for key in histogram.keys():
+                    output.append("<tr><td>{key}</td><td>{value}</td></tr>"
+                                  .format(key=key, value=histogram[key]))
+
+                output.append("\n\t</table>\n</body>\n</html>")
+
+                with open(path + ".html", mode="w+") as f:
+                    x = "".join(output)
+
+                    if f.write(x) > 0:
+                        f.close()
```

Comment:
When you use a `with` statement, you should not bother about closing this context (in this example a file).
```
with locked(myLock):
    # Code here executes with myLock held.  The lock is
    # guaranteed to be released when the block is left (even
    # if via return or by an uncaught exception).
```
taken from: https://www.python.org/dev/peps/pep-0343/